plugins/pluginmanagers/lazy.nix

```diff
@@ -1,222 +1,204 @@
+                  When false, git submodules will not be fetched.
+                  Defaults to true
+                '';
+
+                event =
+
+                  mkNullOrOption (maybeRaw (
+                    either str (listOf str)
+                  )) "Lazy-load on event. Events can be specified as BufEnter or with a pattern like BufEnter *.lua";
+
+                cmd = mkNullOrOption (maybeRaw (either str (listOf str))) "Lazy-load on command";
+
+                ft = mkNullOrOption (maybeRaw (either str (listOf str))) "Lazy-load on filetype";
+
+                keys = mkNullOrOption (maybeRaw (either str (listOf str))) "Lazy-load on key mapping";
+
+                module = mkNullOrOption (enum [ false ]) ''
+                  Do not automatically load this Lua module when it's required somewhere
+                '';
+
+                priority = mkNullOrOption number ''
+                  Only useful for start plugins (lazy=false) to force loading certain plugins first.
+                  Default priority is 50. It's recommended to set this to a high number for colorschemes.
+                '';
+
+                optional = defaultNullOpts.mkBool false ''
+                  When a spec is tagged optional, it will only be included in the final spec,
+                  when the same plugin has been specified at least once somewhere else without optional.
+                  This is mainly useful for Neovim distros, to allow setting options on plugins that may/may not be part
+                  of the user's plugins
+                '';
+
+                opts =
+
+                  mkNullOrOption (maybeRaw (attrsOf anything)) ''
+                    opts should be a table (will be merged with parent specs),
+                    return a table (replaces parent specs) or should change a table.
+                    The table will be passed to the Plugin.config() function.
+                    Setting this value will imply Plugin.config()
+                  '';
+              };
+
+              config.name = lib.mkIf (cfg.pkg != null) (lib.mkDefault "${lib.getName cfg.pkg}");
+              config.dir = lib.mkIf (cfg.pkg != null) (lib.mkDefault "${cfg.pkg}");
+            }
+          )
+        );
+
+      lazyPluginsListType = types.listOf lazyPluginType;
+    in
+    {
+      gitPackage = lib.mkPackageOption pkgs "git" { nullable = true; };
+      luarocksPackage = lib.mkPackageOption pkgs "luarocks" { nullable = true; };
+
+      plugins = lib.mkOption {
+        type = lazyPluginsListType;
+        default = [ ];
+        description = "List of plugins";
+      };
+    };
+
+  extraConfig =
+    cfg:
+    let
+      # The `pkg` property isn't a part of the `lazy.nvim` plugin spec, while
+      # it shouldn't do any harm but it does take up unecessary space in the
+      # init.lua file. Since we're done using it we will strip it from the
+      # final list of specs.
+      removePkgAttrFromPlugin =
```

Comment:
Have you double checked `nixvim-print-init`? By default `toLuaObject` filters out any null or empty items.

Yes, this is a horrible hack, but it's currently the only reasonable way we have to filter out null sub-options from freeform submodules...

Reply:
I'll have a look at the actual code a bit later, but it sounds like we need a `builtins.removeAttrs plugin [ "pkg" ]` somewhere when we convert `cfg.plugins` -> `settings.spec`.

Reply:
My view is (unlike `settings`) `plugins` isn't aiming to be a 1-to-1 mapping of the upstream config format.

Instead it is an alternative (nicer) way to define `settings.spec`, whatever tranformations need to be done to it should be done when defining `plugins.lazy.settings.spec`.

This means users will have a predictable value at `config.plugins.lazy.plugins` if they want to read that in their own modules for any reason.

### Dumb exercise in what's possible

To be clear, I'm **not** recommend you use `apply` here.

But for the sake argument, if you did want to transform the value at `config.plugins.plugins` to be different in some way from what users define, this could be done with either a custom option-type's `merge` function (this is how `coercedTo` works) or defining an `apply` function when calling `mkOption`.

`apply` acts as a final transformation on the fully merged option value, so if I wanted to I could do something really stupid like:

```nix
{ config, lib, ... }:
{
  # Declare foo as a boolean option, that converts its final value to a stupid string...
  options.foo = lib.mkOption {
    type = types.bool;
    apply = v: "was defined as ${lib.boolToString v}!";
  };

  config = {
    # Define foo to be true
    foo = true;

    assertions = [
      {
        # NOTE: while `foo` is defined as a bool, its value is a string!
        # DO NOT DO THIS lol
        assertion = config.foo == "was defined as true!";
        message = "apply failed";
      }
    ];
  };
}
```